### ISSUE TITLE
YOLOv9-t BFLOPs discrepancy

### ISSUE THREAD

**Issue by dbacea** [WongKinYiu/yolov9#506]
In the official table and research paper, the listed number of parameters and FLOPs for YOLOv9-t equal 2.0M and 7.7G. But when training from scratch:
`python -m torch.distributed.launch --nproc_per_node 2 --master_port 9527 train_dual.py --workers 4 --device 0,1 --sync-bn --batch 32 --data data/coco.yaml --img 640 --cfg models/detect/yolov9-t.yaml --weights '' --name yolov9-t --hyp hyp.scratch-high.yaml --min-items 0 --epochs 500 --close-mosaic 15
`

The number of parameters and the BFLOPs printed at the end of training differ from the ones in the official table. The obtained number of parameters equals 3.67M, while the FLOPs equals 16.2G.

Where could this big difference coming from?

**Comment by ankandrew** [comment]
The YOLOv9-T that appears in the table is the converted one (aux branch is removed). The aux branch won't be used during inference, just during training. For a fair comparison, you need to compare the YOLOv9-* (converted) with the corresponding Gelan-* network architecture. The one you see being reported seems to include the aux branch, thus more parameters and FLOPS being shown.

**Comment by dbacea** [comment]
If I understand correctly, the number of parameters and the FLOPs of Gelan-t should be comparable to the ones posted in the official table for YOLOv9-t. 
I've trained a gelan-t from scratch:
`python -m torch.distributed.launch --nproc_per_node 2 --master_port 9527 train.py --workers 4 --device 0,1 --sync-bn --batch 32 --data data/coco.yaml --img 640 --cfg models/detect/gelan-t.yaml --weights '' --name gelan-t --hyp hyp.scratch-high.yaml --min-items 0 --epochs 1 --close-mosaic 15
`

The architecture:
![Annotation 2024-06-20 161856_gelan_t](https://github.com/WongKinYiu/yolov9/assets/60096583/950743f7-55d9-4b2d-a5a7-19054537717b)

Which shows that the number of parameters equals 2442640 and the FLOPs 10.1G.

After training (and layers fusion) the number of parameters equals 2407632 and the FLOPs  9.8G, which still are higher than the ones posted for YOLOv9-t.

![Annotation 2024-06-20 161856_gelan_t_post_training](https://github.com/WongKinYiu/yolov9/assets/60096583/9a333ddd-529b-4988-b62a-d2bc30541030)

Is there an additional step to be taken?

**Comment by ankandrew** [comment]
> If I understand correctly, the number of parameters and the FLOPs of Gelan-t should be comparable to the ones posted in the official table for YOLOv9-t.

Yes. But in the OG post you compared to the `3.67M` number, which corresponds to the one with the aux branch. But yes, `YOLOv9-t-converted` and `Gelan-t`  numbers don't seem to match with what is in the docs table.

> After training (and layers fusion) the number of parameters equals 2407632 and the FLOPs 9.8G, which still are higher than the ones posted for YOLOv9-t.

At this point your question converges to the same as mine, see https://github.com/WongKinYiu/yolov9/issues/505#issue-2363375428. I think the used `gelan-t.yaml` to parametrize the `yolov9-t.yaml` is slightly different than the one provided in the configs.

> Is there an additional step to be taken?

I'm on the same boat as you, these don't seem to match. Let's wait for author answer.

**Comment by ustbzz** [comment]
> 表中显示的 YOLOv9-T 是转换后的 YOLOv9-T（删除了 aux 分支）。aux 分支不会在推理期间使用，只会在训练期间使用。为了公平地进行比较，您需要将 YOLOv9-*（转换后）与相应的 Gelan-* 网络架构进行比较。您看到的被报告的那个似乎包括 aux 分支，因此显示了更多的参数和 FLOPS。

I also encountered a similar problem, and when I tried to reparameterization the t-model trained with the aux network using the reparameterization. ipynb file provided by the author, I found that there was no processing method for the t-model inside; And when I downloaded the pre training weights, I found that the yolov9-t provided by the author was a converted version. Does this mean that the pre training weights have been re parameterized (just my guess)
